File: 2025_07/leetcode2025-07-22.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

class Solution {
public:
    int maximumUniqueSubarray(vector<int>& nums) {
        int n = nums.size();
        int r = 0;
        int curSum = 0;
        int res = 0;
        unordered_map<int, int> cnt;
        for (int l=0; l<n; l++){
            while (r<n && cnt[nums[r]] == 0){
                cnt[nums[r]]++;
                curSum += nums[r];
                r++;
            }
            res = max(res, curSum);
            cnt[nums[l]]--;
            curSum -= nums[l];
        }
        return res;
    }
};

int main(){
    Solution s;
    vector<int> nums = {4,2,4,5,6};
    cout << s.maximumUniqueSubarray(nums) << endl;
}

File: 2025_07/leetcode2025-07-22.py
from typing import List
from collections import defaultdict

class Solution:
    def maximumUniqueSubarray(self, nums: List[int]) -> int:
        # 不定长滑动窗口，只要右端点指向的数字还未出现就将其加入，curSum用来维护当前窗口内数字之和，cnt用来维护当前窗口内每一个数字的出现次数，出现次数为0即代表未出现
        n = len(nums)
        r = 0
        curSum = 0
        res = 0
        cnt = defaultdict(int)
        for l in range(n):
            while r<n and cnt[nums[r]] == 0:
                cnt[nums[r]] += 1
                curSum += nums[r]
                r += 1
            res = max(res, curSum)
            cnt[nums[l]] -= 1
            curSum -= nums[l]
        return res

if __name__ == "__main__":
    s = Solution()
    nums = [4,2,4,5,6]
    print(s.maximumUniqueSubarray(nums))
